File: solution/290/solution.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        strdict = {}
        patterndict = {}
        str += ' '
        temp = ''
        p = 0
        for v in str:
            if v == ' ':
                if p == len(pattern):
                    return False
                if strdict.has_key(temp):
                    if strdict[temp] != pattern[p]:
                        return False
                else:
                    if patterndict.has_key(pattern[p]):
                        return False
                    patterndict[pattern[p]] = 1
                    strdict[temp] = pattern[p]
                p += 1
                temp = ''
            else:
                temp += v
        if p != len(pattern):
            return False
        return True
                    

File: solution/290/solution.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        map<string, char> strdict;
        int patterndict[128] = {0};
        int p = 0;
        string word;
        str.push_back(' ');
        for(int i=0;i < str.length();i++){
            if(str[i]==' '){
                if(p == pattern.length())return false;
                if(strdict.find(word)==strdict.end()){
                    if(patterndict[pattern[p]]!=0)return false;
                    strdict[word] = pattern[p];
                    patterndict[pattern[p]] = 1;
                }else{
                    if(pattern[p]!=strdict[word])return false;
                }
                word = "";
                p ++;
            }else{
                word.push_back(str[i]);
            }
        }
        if(p!=pattern.length()){
            return false;
        }
        return true;
    }
};